File: src/notify.test.ts
import { describe, expect, test } from 'bun:test';
import { mkdtempSync, readFileSync, rmSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { buildNotifyContext, filterHumanChanges, readCursor, writeCursor } from './notify';
import type { ChangeFeedEntry } from './store';

function entry(partial: Partial<ChangeFeedEntry>): ChangeFeedEntry {
  return {
    id: 1,
    entity: 'todo',
    entityId: 'abcd1234',
    actor: 'logan',
    action: 'update',
    at: '2026-07-23T10:00:00.000Z',
    title: '제목',
    boardKey: 'rocky',
    ...partial,
  };
}

describe('filterHumanChanges', () => {
  test('drops agent actors, keeps human actors', () => {
    const entries = [
      entry({ id: 1, actor: 'claude-code' }),
      entry({ id: 2, actor: 'logan' }),
      entry({ id: 3, actor: 'codex' }),
      entry({ id: 4, actor: 'web' }),
    ];
    expect(filterHumanChanges(entries).map((e) => e.id)).toEqual([2, 4]);
  });
});

describe('buildNotifyContext', () => {
  test('null when no entries', () => {
    expect(buildNotifyContext([])).toBeNull();
  });

  test('formats compact korean lines with board, action, and diff', () => {
    const context = buildNotifyContext([
      entry({ action: 'update', changes: { title: ['a', 'b'] } }),
      entry({ id: 2, entity: 'note', action: 'create', title: '메모', boardKey: undefined }),
      entry({ id: 3, action: 'done', title: '끝난 일' }),
    ]);
    expect(context).toContain('rocky-todo');
    expect(context).toContain('[rocky]');
    expect(context).toContain('logan');
    expect(context).toContain('제목');
    expect(context).toContain('title: a → b');
    expect(context).toContain('메모');
    expect(context).toContain('완료');
  });
});

describe('comment lines', () => {
  test('renders a comment with its body instead of a field diff', () => {
    const context = buildNotifyContext([
      entry({
        action: 'comment',
        changes: { comment: [null, '이거 SSE 로도 흘러가나?'] },
        title: '댓글 기능 추가',
      }),
    ]);
    expect(context).toContain('"댓글 기능 추가" 댓글 · "이거 SSE 로도 흘러가나?"');
    expect(context).not.toContain('comment:');
  });

  test('renders an edited comment with the new body', () => {
    const context = buildNotifyContext([
      entry({ action: 'comment-edit', changes: { comment: ['오타', '고침'] } }),
    ]);
    expect(context).toContain('댓글 수정 · "고침"');
  });

  test('folds newlines and truncates a long body', () => {
    const body = `${'가'.repeat(250)}\n둘째 줄`;
    const context = buildNotifyContext([
      entry({ action: 'comment', changes: { comment: [null, body] } }),
    ]);
    expect(context).toContain('…');
    expect(context).not.toContain('\n둘째 줄');
    const line = (context ?? '').split('\n').find((l) => l.includes('댓글')) ?? '';
    expect(line.length).toBeLessThan(300);
  });

  test('agent comments are filtered out before formatting', () => {
    const entries = [
      entry({ id: 1, actor: 'claude-code', action: 'comment', changes: { comment: [null, '봇'] } }),
      entry({ id: 2, actor: 'logan', action: 'comment', changes: { comment: [null, '사람'] } }),
    ];
    const context = buildNotifyContext(filterHumanChanges(entries));
    expect(context).toContain('"사람"');
    expect(context).not.toContain('"봇"');
  });
});

describe('cursor store', () => {
  test('read missing → undefined; write then read round-trips', () => {
    const dir = mkdtempSync(join(tmpdir(), 'rocky-todo-cursor-'));
    const file = join(dir, 'hook-cursors.json');
    try {
      expect(readCursor(file, 'sess-1')).toBeUndefined();
      writeCursor(file, 'sess-1', 42);
      expect(readCursor(file, 'sess-1')).toBe(42);
      writeCursor(file, 'sess-1', 50);
      expect(readCursor(file, 'sess-1')).toBe(50);
      expect(readCursor(file, 'sess-2')).toBeUndefined();
    } finally {
      rmSync(dir, { recursive: true, force: true });
    }
  });

  test('prunes to the most recent 100 sessions', () => {
    const dir = mkdtempSync(join(tmpdir(), 'rocky-todo-cursor-'));
    const file = join(dir, 'hook-cursors.json');
    try {
      for (let i = 0; i < 120; i++) {
        writeCursor(file, `sess-${i}`, i);
      }
      const raw = JSON.parse(readFileSync(file, 'utf8')) as Record<string, unknown>;
      expect(Object.keys(raw).length).toBeLessThanOrEqual(100);
      expect(readCursor(file, 'sess-119')).toBe(119);
      expect(readCursor(file, 'sess-0')).toBeUndefined();

      // 개수만 보면 "어느 100개가 남았는지"를 놓친다 — `at` 이 밀리초라 세션들이 같은
      // 값을 갖기 쉽고, 그때 잘려나가는 구간이 오래된 쪽이 아니라 임의의 밴드가 되던
      // 버그가 있었다(0–11 과 16–23 이 빠지고 12–15 는 남았다). 남은 집합이 정확히
      // 최신 100개인지까지 못 박는다 — 이 단정이 없어서 위 두 줄만으로는 실행마다
      // 통과/실패가 갈렸다.
      const survivors = Array.from({ length: 120 }, (_, i) => i).filter((i) => `sess-${i}` in raw);
      expect(survivors).toEqual(Array.from({ length: 100 }, (_, i) => i + 20));
    } finally {
      rmSync(dir, { recursive: true, force: true });
    }
  });

  test('corrupt cursor file is treated as empty', () => {
    const dir = mkdtempSync(join(tmpdir(), 'rocky-todo-cursor-'));
    const file = join(dir, 'hook-cursors.json');
    try {
      writeCursor(file, 'a', 1);
      require('node:fs').writeFileSync(file, '{broken');
      expect(readCursor(file, 'a')).toBeUndefined();
      writeCursor(file, 'a', 2);
      expect(readCursor(file, 'a')).toBe(2);
    } finally {
      rmSync(dir, { recursive: true, force: true });
    }
  });
});

File: crates/rocky-todo-core/tests/notify_test.rs
//! TS `src/notify.test.ts` 포팅.

use rocky_todo_core::notify::{
    build_notify_context, filter_human_changes, merge_context, read_cursor, write_cursor,
};
use rocky_todo_core::types::{ChangeFeedEntry, Changes, HistoryEntity, HistoryEntry};
use serde_json::json;

fn entry(id: i64, actor: &str, action: &str) -> ChangeFeedEntry {
    ChangeFeedEntry {
        history: HistoryEntry {
            id,
            entity: HistoryEntity::Todo,
            entity_id: "abcd1234".into(),
            actor: actor.into(),
            action: action.into(),
            changes: None,
            at: "2026-07-23T10:00:00.000Z".into(),
        },
        title: "제목".into(),
        board_key: Some("rocky".into()),
    }
}

fn with_changes(mut e: ChangeFeedEntry, changes: serde_json::Value) -> ChangeFeedEntry {
    let map: Changes = changes.as_object().cloned().unwrap_or_default();
    e.history.changes = Some(map);
    e
}

// ── filterHumanChanges ──────────────────────────────────────────────────────

#[test]
fn drops_agent_actors_keeps_human_actors() {
    let entries = vec![
        entry(1, "claude-code", "update"),
        entry(2, "logan", "update"),
        entry(3, "codex", "update"),
        entry(4, "web", "update"),
    ];
    let kept: Vec<i64> = filter_human_changes(entries)
        .iter()
        .map(|e| e.history.id)
        .collect();
    assert_eq!(kept, vec![2, 4]);
}

// ── buildNotifyContext ──────────────────────────────────────────────────────

#[test]
fn none_when_no_entries() {
    assert!(build_notify_context(&[]).is_none());
}

#[test]
fn formats_compact_korean_lines_with_board_action_and_diff() {
    let mut note = entry(2, "logan", "create");
    note.history.entity = HistoryEntity::Note;
    note.title = "메모".into();
    note.board_key = None;
    let mut done = entry(3, "logan", "done");
    done.title = "끝난 일".into();

    let context = build_notify_context(&[
        with_changes(entry(1, "logan", "update"), json!({ "title": ["a", "b"] })),
        note,
        done,
    ])
    .expect("컨텍스트가 있어야 한다");
    for needle in [
        "rocky-todo",
        "[rocky]",
        "logan",
        "제목",
        "title: a → b",
        "메모",
        "완료",
    ] {
        assert!(context.contains(needle), "{needle} 이 없다:\n{context}");
    }
}

// ── 댓글 렌더 ───────────────────────────────────────────────────────────────

#[test]
fn renders_a_comment_with_its_body_instead_of_a_field_diff() {
    let mut e = with_changes(
        entry(1, "logan", "comment"),
        json!({ "comment": [null, "이거 SSE 로도 흘러가나?"] }),
    );
    e.title = "댓글 기능 추가".into();
    let context = build_notify_context(&[e]).unwrap();
    assert!(
        context.contains("\"댓글 기능 추가\" 댓글 · \"이거 SSE 로도 흘러가나?\""),
        "{context}"
    );
    assert!(!context.contains("comment:"), "{context}");
}

#[test]
fn renders_an_edited_comment_with_the_new_body() {
    let e = with_changes(
        entry(1, "logan", "comment-edit"),
        json!({ "comment": ["오타", "고침"] }),
    );
    let context = build_notify_context(&[e]).unwrap();
    assert!(context.contains("댓글 수정 · \"고침\""), "{context}");
}

#[test]
fn folds_newlines_and_truncates_a_long_body() {
    let body = format!("{}\n둘째 줄", "가".repeat(250));
    let e = with_changes(
        entry(1, "logan", "comment"),
        json!({ "comment": [null, body] }),
    );
    let context = build_notify_context(&[e]).unwrap();
    assert!(context.contains('…'), "{context}");
    assert!(!context.contains("\n둘째 줄"), "{context}");
    let line = context.lines().find(|l| l.contains("댓글")).unwrap_or("");
    assert!(line.chars().count() < 300, "{}", line.chars().count());
}

#[test]
fn agent_comments_are_filtered_out_before_formatting() {
    let entries = vec![
        with_changes(
            entry(1, "claude-code", "comment"),
            json!({ "comment": [null, "봇"] }),
        ),
        with_changes(
            entry(2, "logan", "comment"),
            json!({ "comment": [null, "사람"] }),
        ),
    ];
    let context = build_notify_context(&filter_human_changes(entries)).unwrap();
    assert!(context.contains("\"사람\""), "{context}");
    assert!(!context.contains("\"봇\""), "{context}");
}

// ── 커서 저장 ───────────────────────────────────────────────────────────────

#[test]
fn read_missing_is_none_and_write_then_read_round_trips() {
    let dir = tempfile::tempdir().unwrap();
    let file = dir.path().join("hook-cursors.json");
    assert!(read_cursor(&file, "sess-1").is_none());
    write_cursor(&file, "sess-1", 42);
    assert_eq!(read_cursor(&file, "sess-1"), Some(42));
    write_cursor(&file, "sess-1", 50);
    assert_eq!(read_cursor(&file, "sess-1"), Some(50));
    assert!(read_cursor(&file, "sess-2").is_none());
}

/// 개수만 보면 "어느 100개가 남았는지"를 놓친다 — `at` 이 밀리초라 세션들이 같은 값을
/// 갖기 쉽고, 그때 잘려나가는 구간이 오래된 쪽이 아니라 임의의 밴드가 되던 버그가
/// 있었다. 남은 집합이 정확히 최신 100개인지까지 못 박는다.
#[test]
fn prunes_to_the_most_recent_100_sessions() {
    let dir = tempfile::tempdir().unwrap();
    let file = dir.path().join("hook-cursors.json");
    for i in 0..120 {
        write_cursor(&file, &format!("sess-{i}"), i);
    }
    let raw: serde_json::Value =
        serde_json::from_str(&std::fs::read_to_string(&file).unwrap()).unwrap();
    let keys = raw.as_object().unwrap();
    assert!(keys.len() <= 100, "{}", keys.len());
    assert_eq!(read_cursor(&file, "sess-119"), Some(119));
    assert!(read_cursor(&file, "sess-0").is_none());

    let survivors: Vec<i64> = (0..120)
        .filter(|i| keys.contains_key(&format!("sess-{i}")))
        .collect();
    let expected: Vec<i64> = (20..120).collect();
    assert_eq!(survivors, expected);
}

#[test]
fn corrupt_cursor_file_is_treated_as_empty() {
    let dir = tempfile::tempdir().unwrap();
    let file = dir.path().join("hook-cursors.json");
    write_cursor(&file, "a", 1);
    std::fs::write(&file, "{broken").unwrap();
    assert!(read_cursor(&file, "a").is_none());
    write_cursor(&file, "a", 2);
    assert_eq!(read_cursor(&file, "a"), Some(2));
}

// ── mergeContext ────────────────────────────────────────────────────────────

#[test]
fn both_present_are_joined_with_a_blank_line() {
    assert_eq!(
        merge_context(&[Some("A".into()), Some("B".into())]).as_deref(),
        Some("A\n\nB")
    );
}

#[test]
fn a_single_part_is_returned_alone() {
    assert_eq!(
        merge_context(&[None, Some("B".into())]).as_deref(),
        Some("B")
    );
    assert_eq!(
        merge_context(&[Some("A".into()), None]).as_deref(),
        Some("A")
    );
}

#[test]
fn nothing_to_inject_is_none() {
    assert!(merge_context(&[None, None]).is_none());
}

#[test]
fn empty_strings_count_as_absent() {
    assert_eq!(
        merge_context(&[Some(String::new()), Some("B".into())]).as_deref(),
        Some("B")
    );
}
